File: src/main/java/com/algorithm/question1/javarouka/ReverseString.js
function isEmpty(str) {
    return !str || !str.length;
}

exports.default = function reverseStringCenter(str) {

    if(isEmpty(str)) return str;

    const len = str.length;
    const center = len/2;
    let start = 0;
    let back = len;

    const result = [];

    for(;start < center; start++, back--) {
        result[back - 1] = str[start];
        result[start] = str[back - 1];
    }

    return result.join('');
};

exports.reverseStringUseReduce = function(str) {

    if(isEmpty(str)) return str;

    return str.split('')
        .reduce((result, val, index, orig) => {
            result.push(orig[orig.length - index - 1]);
            return result;
        }, [])
        .join('');
};

exports.reverseStringFullLoop = function(str) {

    if(isEmpty(str)) return str;

    const stack = [];
    const fixSize = str.length;
    let len = str.length;

    while(len--) {
        stack[fixSize-len] = str[len];
    }

    return stack.join('')

};

exports.reverseStringCenter = exports.default;

File: src/main/java/com/algorithm/question1/javarouka/ReverseString.java
package com.algorithm.question1.javarouka;

public class StringReverser {

    private static boolean isEmpty(CharSequence str) {
        return str == null || str.length == 0
    }

    public static String reverse(CharSequence str) {

        if(isEmpty(str)) {
            return str;
        }

        // 이 부분에 대한 고려 필요
        // 엄청나게 긴 문자열이 올 경우, 어떻게 대처할 것인가..?
        int len = str.length;
        int center = len/2;
        int start = 0;
        int back = len;

        StringBuilder result = new StringBuilder(len);

        for(;start < center; start++, back--) {
            result.insert(back - 1, str.charAt(start));
            result.insert(start, str.charAt(back - 1));
        }

        return result.toString();

    }

}